tests/unit/ip_router_interface/v0/test_ip_router.py

```diff
@@ -0,0 +1,533 @@
+            }
+        ]
+        harness.update_relation_data(
+            rel_1_id, "ip-router-provider-a", {"networks": json.dumps(existing_network_1)}
+        )
+        # Create network 2
+        existing_network_2 = [
+            {"network": "192.168.252.0/24", "gateway": "192.168.252.1"},
+            {"network": "192.168.251.0/24", "gateway": "192.168.251.1/24"},
+        ]
+        harness.update_relation_data(
+            rel_2_id, "ip-router-provider-b", {"networks": json.dumps(existing_network_2)}
+        )
+        harness.charm.RouterRequirer.relationship_name = IP_ROUTER_REQUIRER_RELATION_NAME
+        assert (
+            harness.charm.RouterRequirer.get_all_networks()
+            == existing_network_1 + existing_network_2
+        )
+
+    def test_request_new_network(self):
+        harness = self._setup()
+
+        # Create a relation
+        rel_id = harness.add_relation(IP_ROUTER_REQUIRER_RELATION_NAME, "ip-router-provider")
+        harness.add_relation_unit(rel_id, "ip-router-provider/0")
+
+        # Request a network
+        network_request = [
+            {
+                "network": "192.168.252.0/24",
+                "gateway": "192.168.252.1",
+            }
+        ]
+        harness.charm.RouterRequirer.relationship_name = IP_ROUTER_REQUIRER_RELATION_NAME
+        harness.charm.RouterRequirer.request_network(network_request)
+
+        assert harness.get_relation_data(rel_id, "ip-router-requirer") == {
+            "networks": json.dumps(network_request),
+            "network-name": IP_ROUTER_REQUIRER_RELATION_NAME,
+        }
+
+    def test_request_new_network_multiple(self):
+        harness = self._setup()
+
+        # Create a relation
+        rel_id = harness.add_relation(IP_ROUTER_REQUIRER_RELATION_NAME, "ip-router-provider")
+        harness.add_relation_unit(rel_id, "ip-router-provider/0")
+
+        # Request a network
+        network_request = [
+            {"network": "192.168.252.0/24", "gateway": "192.168.252.1"},
+            {"network": "192.168.250.0/24", "gateway": "192.168.250.1"},
+        ]
+        harness.charm.RouterRequirer.relationship_name = IP_ROUTER_REQUIRER_RELATION_NAME
+        harness.charm.RouterRequirer.request_network(network_request)
+
+        assert harness.get_relation_data(rel_id, "ip-router-requirer") == {
+            "networks": json.dumps(network_request),
+            "network-name": IP_ROUTER_REQUIRER_RELATION_NAME,
+        }
+
+    def test_request_new_network_missing_relation(self):
+        harness = self._setup()
+
+        # Request a network
+        network_request = [
+            {
+                "network": "192.168.252.0/24",
+                "gateway": "192.168.252.1",
+            }
+        ]
+        harness.charm.RouterRequirer.relationship_name = IP_ROUTER_REQUIRER_RELATION_NAME
+        try:
+            harness.charm.RouterRequirer.request_network(network_request)
```

Comment:
I would use the `with self.assertRaises(<ErrorType>):` structure, to assert that the exception actually happens when we are expecting it. If something changes on the code and the exception is no longer raised for any reason, the test won't fail.
This comment applies for all the test that have a `try-except` on them.

Reply:
Done